scalanet/src/io/iohk/scalanet/peergroup/dynamictls/DynamicTLSPeerGroup.scala

```diff
@@ -0,0 +1,395 @@
+      case NonFatal(e) => Task.raiseError(InitializationError(e.getMessage, e.getCause))
+    }
+
+  override def processAddress: PeerInfo = config.peerInfo
+
+  override def client(to: PeerInfo): Task[Channel[PeerInfo, M]] = {
+    // Creating new ssl context for each client is necessary, as this is only reliable way to pass peerInfo to TrustManager
+    // which takes care of validating certificates and server node id.
+    // Using Netty SSLEngine.getSession.putValue does not work as expected as until successfulhandshake there is no separate
+    // session for each connection.
+    new ClientChannelImpl[M](
+      to,
+      clientBootstrap,
+      DynamicTLSPeerGroupUtils.buildCustomSSlContext(SSLContextForClient(to), config),
+      codec
+    ).initialize
+  }
+
+  override def server(): ConnectableObservable[ServerEvent[PeerInfo, M]] = serverSubject
+
+  override def shutdown(): Task[Unit] = {
+    serverSubject.onComplete()
+    for {
+      _ <- toTask(serverBind.channel().close())
+      _ <- toTask(workerGroup.shutdownGracefully())
+    } yield ()
+  }
+}
+
+object DynamicTLSPeerGroup {
+  case class PeerInfo(id: BitVector, address: InetMultiAddress)
+  object PeerInfo {
+    implicit val peerInInst = new Addressable[PeerInfo] {
+      override def getAddress(a: PeerInfo): InetSocketAddress = a.address.inetSocketAddress
+    }
+  }
+
+  case class Config(
+      bindAddress: InetSocketAddress,
+      peerInfo: PeerInfo,
+      connectionKeyPair: KeyPair,
+      connectionCertificate: X509Certificate
+  )
+
+  object Config {
+    // FIXME: For now we support only Secp256 keys in ethereum format
+    def apply(
+        bindAddress: InetSocketAddress,
+        keyType: KeyType,
+        hostKeyPair: KeyPair,
+        secureRandom: SecureRandom
+    ): Try[Config] = {
+
+      SignedKeyExtensionNodeData(keyType, hostKeyPair, Secp256r1, secureRandom).map { nodeData =>
+        Config(
+          bindAddress,
+          PeerInfo(nodeData.calculatedNodeId, InetMultiAddress(bindAddress)),
+          nodeData.generatedConnectionKey,
+          nodeData.certWithExtension
+        )
+      }
+    }
+
+    def apply(
+        bindAddress: InetSocketAddress,
+        keyType: KeyType,
+        hostKeyPair: AsymmetricCipherKeyPair,
+        secureRandom: SecureRandom
+    ): Try[Config] = {
+      val convertedKeyPair = CryptoUtils.convertBcToJceKeyPair(hostKeyPair)
+      Config(bindAddress, keyType, convertedKeyPair, secureRandom)
+    }
+  }
+
+  private class MessageNotifier[M](
+      val messageSubject: ConnectableSubject[M],
+      codec: StreamCodec[M]
+  ) extends ChannelInboundHandlerAdapter {
+
+    private val log = LoggerFactory.getLogger(getClass)
+
+    override def channelInactive(channelHandlerContext: ChannelHandlerContext): Unit =
+      messageSubject.onComplete()
+
+    override def channelRead(ctx: ChannelHandlerContext, msg: Any): Unit = {
+      val byteBuf = msg.asInstanceOf[ByteBuf]
```

Comment:
is it safe to do casting before try ?

Reply:
yup ale netty messages read from channels are bytebuffers at this level. Also we are casting ot most general bytebuffer type possible so even if implementation would change we should be safe here.